Persist header and toolbar toggle status in nbconfig

gnestor opened the pull request:
A first stab at https://github.com/jupyter/notebook/issues/1661

gnestor commented:
@Carreau I am following your lead from https://github.com/jupyter/notebook/pull/1423/commits/c04046430d2b10923ae3d08d367b629351283f1d. However, neither the `header` or `toolbar` properties are being persisted in `notebook.config.data` after refresh. I don't fully understand how nbconfig works so I would appreciate your input here (I've enabled "Allow edits from maintainers" so you should be able to push to this PR).

Carreau commented:
Yeah, nbconfig thing is complex. let me have a look. Thanks for that.

Carreau commented:
So it seem you get the config right, the problem is the asynchronicity and the config return defaults before being completly loaded. 

if you `cat ~/.jupyter/nbconfig/notebook.json` you can see the values get changed. 

You might want to add a `this.config.loaded.then(function(){stuff(that.config.data)}`

The `resolve`  part of the promise does not get passed the new config apparently, but that should be fixable in `notebook/services/config.js` around line 34:

```
 ConfigSection.prototype._load_done = function() {
        if (!this._one_load_finished) {
            this._one_load_finished = true;
            this._finish_firstload();  // might want to pass self.data here. 
        }
    };
```

Does that make sens ?

takluyver commented:
The idea with the config is that you can create a `ConfigWithDefaults` object, and then you have a choice of:

```javascript
// Get a value immediately, using the defaults if real values aren't loaded yet
visible = c.get_sync('toolbar_visible');

// Wait for the value if necessary
c.get('toolbar_visible').then(function(visible) {...});
```

gnestor commented:
@Carreau Your `this.config.loaded.then()` suggestion worked great...

Two questions:

1. Is there a place to define default config values so that I don't have to wonder if `this.config.data.Header` is defined or not?
2. Is there a better place to check the config values and hide/show the header/toolbar than in the initialization of the `Notebook` class (e.g. `Notebook.load_notebook`)?

takluyver commented:
> Is there a place to define default config values so that I don't have to wonder if this.config.data.Header is defined or not?

There's a `ConfigWithDefaults` class, used like this:

```javascript
configwd = ConfigWithDefaults(this.config, {
    header: true,
    toolbar: true,
});
configwd.get('header').then(function(header) {
    if (!header) {
        $('#header-container').hide();
        ...
    }
});
```

gnestor commented:
Thanks @takluyver! Ok I switched from `ConfigSection` to `ConfigWithDefaults` to take advantage of defaults. I also moved header/toolbar toggle logic from `actions.js` to `notebook.js` to be consistent with the line numbers toggle implementation. 

I think this is ready for review...

gnestor commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -164,29 +164,79 @@ import {ShortcutEditor} from 'notebook/js/shortcuteditor';
+            }
+        });
+        
+        this.class_config.get('Header').then(function(header) {
```
To revisit my other question: Is there a better place to put this (like `Notebook.load_notebook`)?

gnestor commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -164,29 +164,79 @@ import {ShortcutEditor} from 'notebook/js/shortcuteditor';
+        
+        Object.defineProperty(this, 'header', {
+            get: function() {
+                return that.class_config.get_sync('Header');
```
I switched from ConfigSection to ConfigWithDefaults to take advantage of defaults.

gnestor commented on notebook/static/notebook/js/actions.js:
```diff
@@ -462,38 +462,23 @@ define(function(require){
+        'toggle-header':{
+            help: 'hide/show the header',
             handler : function(env){
-                $('div#maintoolbar').toggle();
```
I also moved header/toolbar toggle logic from actions.js to notebook.js to be consistent with the line numbers toggle implementation.

gnestor commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -164,29 +164,79 @@ import {ShortcutEditor} from 'notebook/js/shortcuteditor';
-          }
-          
-        })
+            get: function() {
```
I tried to update the line numbers toggle logic to use `ConfigWithDefaults` (via `that.class_config.get_sync` and `that.class_config.set`) to be consistent with header and toolbar but it looks like the line numbers toggle status is being read from `notebook.config.data` regardless. `notebook.json` looks like:

```json
{
  "Notebook": {
    "Toolbar": false,
    "Cell": {
      "cm_config": {
        "lineNumbers": false
      }
    },
    "Header": false
  },
  "Cell": {
    "cm_config": {
      "lineNumbers": true
    }
  }
}
```

The top-level `Cell` is set using `notebook.config.update` while `Notebook.Cell` is set using `notebook.class_config.set`. Changing the value of `Notebook.Cell.cm_config.lineNumbers` has no effect even after changing the line numbers toggle logic, so the cells/Codemirror appear to be accessing `Cell.cm_config.lineNumbers` directly vs. accessing `notebook.lineNumbers`. I know this is probably confusing and that's because it is! It doesn't really matter because everything is working but it would be nice if these 3 properties shared the same implementation.

takluyver commented on notebook/static/notebook/js/actions.js:
```diff
@@ -462,38 +462,23 @@ define(function(require){
                 env.notebook.show_command_palette();
             }
         },
-        'show-all-line-numbers': {
```
Actions are quite public API - they're used to define keyboard shortcuts - so I'd rather not remove them without a good reason.

gnestor replied:
Good point. I added back the show/hide line numbers as well as adding show/hide actions for header and toolbar.

gnestor commented on notebook/static/notebook/js/notebook.js:
```diff
@@ -164,29 +164,79 @@ import {ShortcutEditor} from 'notebook/js/shortcuteditor';
+        
+        this.class_config.get('Header').then(function(header) {
+            if (header === false) {
+                that.header = false;
```
This is a good point, thanks for the feedback. I just pushed a new commit that moves the toggle logic to the actions and then calls the actions to toggle the header/toolbar on notebook load.

gnestor commented:
@Carreau Would you care to review? It's working for me and I think it's quite a bit cleaner given @takluyver's feedback. Two questions for you:

1. I would like `notebook.line_numbers` to use `ConfigWithDefaults` as well to keep these properties consistent. I tried it and the issue is this: `ConfigWithDefaults` (or `notebook.class_config.set`) saves the line numbers toggle state at `Notebook.Cell.cm_config.lineNumbers` and `ConfigSection` (or `notebook.config.update`) saves it at `Cell.cm_config.lineNumbers`. After switching to `ConfigWithDefaults`, the cells no longer reflected the state in `Notebook.Cell.cm_config.lineNumbers` after refresh. However, if I manually changed the value of `Cell.cm_config.lineNumbers` in `notebook.json`, then they would reflect that. **Sooo, the cells are not accessing the line numbers toggle state from `notebook.line_numbers` but appear to accessing it via `notebook.config.data.Cell.cm_config.lineNumbers`. My question for you: where does Codemirror look to determine whether a cell should display line numbers or not (specifically on load)?**
2. Is there a better place to put the "check nbconfig for header/toolbar toggle state and hide/show them accordingly" behavior than in the initialization of the `notebook` class (like `notebook.load_notebook` for example)?

Carreau commented:
I'm giving a seminar today and will talk to a lot of people before and
after so won't be online a lot. I'll have a look ASAP.

On Thu, Sep 22, 2016 at 12:54 PM, Grant Nestor <notifications@github.com>
wrote:

> @Carreau <https://github.com/Carreau> Would you care to review? It's
> working for me and I think it's quite a bit cleaner given @takluyver
> <https://github.com/takluyver>'s feedback. Two questions for you:
>
>    1. I would like notebook.line_numbers to use ConfigWithDefaults as
>    well to keep these properties consistent. I tried it and the issue is this:
>    ConfigWithDefaults (or notebook.class_config.set) saves the line
>    numbers toggle state at Notebook.Cell.cm_config.lineNumbers and
>    ConfigSection (or notebook.config.update) saves it at
>    Cell.cm_config.lineNumbers. After switching to ConfigWithDefaults, the
>    cells no longer reflected the state in Notebook.Cell.cm_config.
>    lineNumbers after refresh. However, if I manually changed the value of
>    Cell.cm_config.lineNumbers in notebook.json, then they would reflect
>    that. *Sooo, the cells are not accessing the line numbers toggle state
>    from notebook.line_numbers but appear to accessing it via
>    notebook.config.data.Cell.cm_config.lineNumbers. My question for you: where
>    does Codemirror look to determine whether a cell should display line
>    numbers or not (specifically on load)?*
>    2. Is there a better place to put the "check nbconfig for
>    header/toolbar toggle state and hide/show them accordingly" behavior than
>    in the initialization of the notebook class (like
>    notebook.load_notebook for example)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/notebook/pull/1769#issuecomment-249010144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUez2t4sDZ71KaM56Al57PNjhchdlrxks5qstzogaJpZM4J_FhF>
> .
>

gnestor commented:
@Carreau Ok no problem!

Carreau commented:
> Notebook.Cell.cm_config.lineNumbers

this is due to the fact that line number could be toggle per-cell. 
I think this is/was a mistake, and I'm fine not storing this config there anymore.

Currently here is how it seem to work...:

Codemirror instances are created on the `create_element` methods of `codecell.js`, and `textcell.js` (` this.code_mirror = new CodeMirror(input_area.get(0), this._options.cm_config);` lines.

`this._options.cm_config` contain likely the `lineNumber` field which is set on the parent class `Cell` (from `cell.js`) generated on the line `this._options = utils.mergeopt({}, cell_config_data, class_config_data);`, and `class_config_data` get the config synchronously `class_config_data = this.class_config.get_sync();`.

Does that make (some) sens ?

gnestor commented:
Yep, it makes perfect sense. The culprit was https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/notebook.js#L1185. I ran into a little error but it's almost there...

gnestor commented:
@Carreau Let's merge this and I will work off of another branch to try to get `notebook.line_numbers` to use `ConfigWithDefaults`.

Carreau commented:
You are right, let's try ! Thanks !
